File: algo/src/main/java/com/hartwig/actin/algo/evaluation/tumor/HasKnownCnsMetastases.java
package com.hartwig.actin.algo.evaluation.tumor;

import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.algo.datamodel.Evaluation;
import com.hartwig.actin.algo.datamodel.EvaluationResult;
import com.hartwig.actin.algo.datamodel.ImmutableEvaluation;
import com.hartwig.actin.algo.evaluation.EvaluationFactory;
import com.hartwig.actin.algo.evaluation.EvaluationFunction;

import org.jetbrains.annotations.NotNull;

public class HasKnownCnsMetastases implements EvaluationFunction {

    HasKnownCnsMetastases() {
    }

    @NotNull
    @Override
    public Evaluation evaluate(@NotNull PatientRecord record) {
        Boolean hasCnsLesions = record.clinical().tumor().hasCnsLesions();
        Boolean hasBrainLesions = record.clinical().tumor().hasBrainLesions();

        if (hasCnsLesions == null && hasBrainLesions == null) {
            return EvaluationFactory.unrecoverable()
                    .result(EvaluationResult.FAIL)
                    .addFailSpecificMessages("Data regarding presence of CNS metastases is missing, assuming there are none")
                    .addFailGeneralMessages("Assuming no known CNS metastases")
                    .build();
        }

        boolean hasKnownCnsMetastases = hasCnsLesions != null && hasCnsLesions;

        boolean hasAtLeastActiveBrainMetastases = false;
        if (hasBrainLesions != null && hasBrainLesions) {
            hasKnownCnsMetastases = true;
            hasAtLeastActiveBrainMetastases = true;
        }

        EvaluationResult result = hasKnownCnsMetastases ? EvaluationResult.PASS : EvaluationResult.FAIL;

        ImmutableEvaluation.Builder builder = EvaluationFactory.unrecoverable().result(result);
        if (result == EvaluationResult.FAIL) {
            builder.addFailSpecificMessages("No known CNS metastases present");
            builder.addFailGeneralMessages("No known CNS metastases");
        } else if (result == EvaluationResult.PASS) {
            if (hasAtLeastActiveBrainMetastases) {
                builder.addPassSpecificMessages("Brain metastases are present, these are considered CNS metastases");
                builder.addPassGeneralMessages("Brain (CNS) metastases");
            } else {
                builder.addPassSpecificMessages("CNS metastases are present");
                builder.addPassGeneralMessages("CNS metastases");
            }
        }

        return builder.build();
    }
}

File: algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/tumor/HasKnownCnsMetastases.kt
package com.hartwig.actin.algo.evaluation.tumor

import com.hartwig.actin.algo.evaluation.EvaluationFactory
import com.hartwig.actin.algo.evaluation.EvaluationFunction
import com.hartwig.actin.datamodel.PatientRecord
import com.hartwig.actin.datamodel.algo.Evaluation

class HasKnownCnsMetastases : EvaluationFunction {

    override fun evaluate(record: PatientRecord): Evaluation {
        with(record.tumor) {

            return when {
                hasCnsLesions == true -> {
                    EvaluationFactory.pass("CNS metastases in provided lesions")
                }

                hasBrainLesions == true -> {
                    EvaluationFactory.pass("Brain metastases in provided lesions")
                }

                hasSuspectedCnsLesions == true || hasSuspectedBrainLesions == true -> {
                    val message = "CNS metastases in provided lesions but suspected lesions only"
                    EvaluationFactory.warn(message)
                }

                hasCnsLesions == null || hasBrainLesions == null -> {
                    val message = "Undetermined if CNS metastases based on provided lesions"
                    EvaluationFactory.undetermined(message)
                }

                else -> EvaluationFactory.fail("No known CNS metastases in provided lesions")
            }
        }
    }
}